### PR TITLE
Object Factory to aid extensibility.

### PR DESCRIPTION
**Pull request by rmclaren**
## Description

There has been a lot of activity lately to add new export variables to the BUFR code. Currently you have to extend an if else statement in exports in order to add their new variable. This PR adds a generic ObjectFactory that can be used in the place of the if else tree's in order to make the extension creation more straight forward. It will also force a certain uniformity to splits, filters, variables and so on. 

## Acceptance Criteria (Definition of Done)

`ctest -R iodaconv_bufr` tests pass

### CODE REVIEW

**Review by srherbener — COMMENTED** [review]
Nice update! Thanks for adding this. I just have a couple minor comments.

File: src/bufr/ObjectFactory.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+#include <memory>
+#include <string>
+#include <ostream>
+
+#include "eckit/exception/Exceptions.h"
+
+
+namespace Ingester
+{
+    /// \brief Factory that is used to create Parsers
```
[the remark this answers is not in the document]
Comment (rmclaren) [reply src/bufr/ObjectFactory.h]: Thanks! I guess I dropped the ball when updating the documentation..

File: src/bufr/ObjectFactory.h [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                return std::make_shared<T>(args...);
+            }
+        };
+
+     public:
+        virtual ~ObjectFactory() = default;
+
+        /// \brief Create a Parser
+        /// \param conf Provides the name of the parser we want to create
+        std::shared_ptr<U> create(const std::string& objectName, Args... args)
```
[the remark this answers is not in the document]
Comment (rmclaren) [reply src/bufr/ObjectFactory.h:63]: I updated the doc strings to explain this a little better. Thanks:)

**Review by srherbener — APPROVED** [review]
Thanks for addressing my comments. I'm approving, but the CI test failures need to be resolved before merging.

**Comment by rmclaren** [comment]
@srherbener For some reason I'm unable to get to the cdash page to see what the issue is... It's telling me access denied.

**Comment by srherbener** [comment]
@rmclaren here is the output from CDASH.
```
test_iodaconv_gsidiag_conv_uv:


DIFFER : VARIABLE : eastward_wind : POSITION : [1] : VALUES : 4.28 <> 4.28
DIFFER : VARIABLE : eastward_wind : POSITION : [60] : VALUES : 3.9 <> 3.9
DIFFER : VARIABLE : eastward_wind : POSITION : [62] : VALUES : 7 <> 7
DIFFER : VARIABLE : eastward_wind : POSITION : [63] : VALUES : 7.455 <> 7.455
DIFFER : VARIABLE : eastward_wind : POSITION : [64] : VALUES : 7.466 <> 7.466
DIFFER : VARIABLE : eastward_wind : POSITION : [65] : VALUES : 7 <> 7
DIFFER : VARIABLE : northward_wind : POSITION : [1] : VALUES : 4.28 <> 4.28
DIFFER : VARIABLE : northward_wind : POSITION : [60] : VALUES : 3.9 <> 3.9
DIFFER : VARIABLE : northward_wind : POSITION : [62] : VALUES : 7 <> 7
DIFFER : VARIABLE : northward_wind : POSITION : [63] : VALUES : 7.455 <> 7.455
DIFFER : VARIABLE : northward_wind : POSITION : [64] : VALUES : 7.466 <> 7.466
DIFFER : VARIABLE : northward_wind : POSITION : [65] : VALUES : 7 <> 7
DIFFER : VARIABLE : eastward_wind : POSITION : [62] : VALUES : 7 <> 7
DIFFER : VARIABLE : eastward_wind : POSITION : [63] : VALUES : 7.455 <> 7.455
DIFFER : VARIABLE : eastward_wind : POSITION : [64] : VALUES : 7.466 <> 7.466
DIFFER : VARIABLE : eastward_wind : POSITION : [65] : VALUES : 7 <> 7
DIFFER : VARIABLE : eastward_wind : POSITION : [68] : VALUES : 6.8792 <> 6.8792
DIFFER : VARIABLE : northward_wind : POSITION : [1] : VALUES : 4.28 <> 4.28
DIFFER : VARIABLE : northward_wind : POSITION : [8] : VALUES : 7.5202 <> 7.5202
DIFFER : VARIABLE : northward_wind : POSITION : [13] : VALUES : 7 <> 7
DIFFER : VARIABLE : northward_wind : POSITION : [14] : VALUES : 6.0168 <> 6.0168
DIFFER : VARIABLE : northward_wind : POSITION : [19] : VALUES : 7 <> 7
DIFFER : VARIABLE : northward_wind : POSITION : [54] : VALUES : 7.388 <> 7.388
DIFFER : VARIABLE : northward_wind : POSITION : [55] : VALUES : 7.476 <> 7.476
DIFFER : VARIABLE : northward_wind : POSITION : [62] : VALUES : 7 <> 7
DIFFER : VARIABLE : northward_wind : POSITION : [63] : VALUES : 7.455 <> 7.455
DIFFER : VARIABLE : northward_wind : POSITION : [64] : VALUES : 7.466 <> 7.466
DIFFER : VARIABLE : northward_wind : POSITION : [65] : VALUES : 7 <> 7
DIFFER : VARIABLE : northward_wind : POSITION : [68] : VALUES : 6.8792 <> 6.8792
DIFFER : VARIABLE : eastward_wind : POSITION : [1] : VALUES : 4.28 <> 4.28
DIFFER : VARIABLE : eastward_wind : POSITION : [60] : VALUES : 3.9 <> 3.9
DIFFER : VARIABLE : eastward_wind : POSITION : [62] : VALUES : 7 <> 7
DIFFER : VARIABLE : eastward_wind : POSITION : [63] : VALUES : 7.455 <> 7.455
DIFFER : VARIABLE : eastward_wind : POSITION : [64] : VALUES : 7.466 <> 7.466
DIFFER : VARIABLE : eastward_wind : POSITION : [65] : VALUES : 7 <> 7
Variable       Group              Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
eastward_wind  /GsiAdjustObsError     6 -7.15256e-07  2.6226e-06 -4.76837e-07 4.76837e-07 9.53674e-07 -1.19209e-07 4.70839e-07
northward_wind /GsiAdjustObsError     6 -7.15256e-07  2.6226e-06 -4.76837e-07 4.76837e-07 9.53674e-07 -1.19209e-07 4.70839e-07
eastward_wind  /GsiFinalObsError      5 -4.76837e-07 2.38419e-06 -4.76837e-07 4.76837e-07 9.53674e-07 -9.53674e-08 5.22349e-07
northward_wind /GsiFinalObsError     12 -1.90735e-06 5.72205e-06 -4.76837e-07 4.76837e-07 9.53674e-07 -1.58946e-07 4.69557e-07
eastward_wind  /GsiInputObsError      6 -7.15256e-07  2.6226e-06 -4.76837e-07 4.76837e-07 9.53674e-07 -1.19209e-07 4.70839e-07
logo


test_iodaconv_gsidiag_rad:


Input Parameters: ObsBias=None QCVars=None TestRefs=None
Satellite radiance obs processed, wrote to: testrun//amsua_aqua_obs_2018041500.nc4
DIFFER : VARIABLE : brightness_temperature : POSITION : [0,7] : VALUES : 0.300003 <> 0.300003
DIFFER : VARIABLE : brightness_temperature : POSITION : [0,9] : VALUES : 0.350032 <> 0.350032
DIFFER : VARIABLE : brightness_temperature : POSITION : [0,10] : VALUES : 0.450182 <> 0.450182
Variable               Group             Count         Sum     AbsSum          Min         Max       Range        Mean      StdDev
brightness_temperature /GsiFinalObsError     3 2.98023e-08 8.9407e-08 -2.98023e-08 2.98023e-08 5.96046e-08 9.93411e-09 3.44128e-08

```

**Comment by rmclaren** [comment]
So I lowered the precision on the gsidiag converter tests that were failing and that seemed to help things. This is not my code however (tests have nothing to do with anything I did)...

**Comment by emilyhcliu** [comment]
@rmclaren Your code change looks good to me. This enhancement simplifies the steps for adding new variables.   Thanks!

**Comment by BenjaminRuston** [comment]
@rmclaren am waiting on another PR to clear but then will focus on this one next.  If passes all CI ready to go and is largely deconflicted with the `sprint` branches correct?

@gthompsnJCSDA similarly was going to update and merge if successful CI testing

**Comment by rmclaren** [comment]
@BenjaminRuston You are correct. This PR shouldn't overlap with anything in the code sprint. It's just a source code change.

**Comment by BenjaminRuston** [comment]
@rmclaren getting a failure on the build:

```
[ 76%] Building CXX object iodaconv/src/bufr/CMakeFiles/ingester.dir/BufrParser/Exports/Export.cpp.o
--
228 | /jcsda/ioda-bundle/iodaconv/src/bufr/BufrParser/Exports/Export.cpp(140): error: identifier "variable" is undefined
229 | variable = std::make_shared<DatetimeVariable>(key, groupByField, dtconf);
230 | ^
231 |  
232 | /jcsda/ioda-bundle/iodaconv/src/bufr/BufrParser/Exports/Export.cpp(145): error: identifier "variable" is undefined
233 | variable = std::make_shared<AircraftAltitudeVariable>(key, groupByField, dtconf);
234 | ^
235 |  
236 | /jcsda/ioda-bundle/iodaconv/src/bufr/BufrParser/Exports/Export.cpp(150): error: identifier "variable" is undefined
237 | variable = std::make_shared<TimeoffsetVariable>(key, groupByField, dtconf);
238 | ^
239 |  
240 | /jcsda/ioda-bundle/iodaconv/src/bufr/BufrParser/Exports/Export.cpp(154): error: name followed by "::" must be a class or namespace name
241 | Transforms transforms = TransformBuilder::makeTransforms(subConf);
242 | ^
243 |  
244 | compilation aborted for /jcsda/ioda-bundle/iodaconv/src/bufr/BufrParser/Exports/Export.cpp (code 2)
```

**Comment by rmclaren** [comment]
@BenjaminRuston Fixed the issue resulting from merging in the develop branch.

**Pull request merged**